sealir-tutorials/ch06_mlir_backend.py

```diff
@@ -0,0 +1,588 @@
+    elif isinstance(mlir_ty, ir.F32Type):
+        return ctypes.pointer(ctypes.c_float(val))
+    elif isinstance(mlir_ty, ir.F64Type):
+        return ctypes.pointer(ctypes.c_double(val))
+
+
+@dataclass(frozen=True)
+class JitCallable:
+    jit_func: Callable
+
+    @classmethod
+    def from_pointer(cls, jit_module, input_types, output_types):
+        engine = execution_engine.ExecutionEngine(jit_module)
+
+        assert (
+            len(output_types) == 1
+        ), "Execution of functions with output arguments > 1 not supported"
+        res_ptr = get_exec_ptr(output_types[0], 0)
+
+        def jit_func(*input_args):
+            assert len(input_args) == len(input_types)
+            for arg, arg_ty in zip(input_args, input_types):
+                # assert isinstance(arg, arg_ty)
+                # TODO: Assert types here
+                pass
+            input_exec_ptrs = [
+                get_exec_ptr(ty, val)
+                for ty, val in zip(input_types, input_args)
+            ]
+            engine.invoke("func", *input_exec_ptrs, res_ptr)
+
+            return res_ptr.contents.value
+
+        return cls(jit_func)
+
+    def __call__(self, *args: Any) -> Any:
+        return self.jit_func(*args)
+
+
+# + [markdown] jp-MarkdownHeadingCollapsed=true
+# Example 1: simple if-else
+# -
+
+
+def example_1(a, b):
+    if a > b:
+        z = a - b
+    else:
+        z = b - a
+    return z + a
+
+
+if __name__ == "__main__":
+    jt = compiler_pipeline(
+        example_1,
+        argtypes=(Int64, Int64),
+        ruleset=(if_else_ruleset | facts_function_types),
+        verbose=True,
+        converter_class=ConditionalExtendGraphtoRVSDG,
+        cost_model=MyCostModel(),
+        backend=Backend(),
+    )
+    args = (10, 33)
+    run_test(example_1, jt, args, verbose=True)
+    args = (7, 3)
+    run_test(example_1, jt, args, verbose=True)
+
+
+# ## Example 2: add `float()`
+
+
+def example_2(a, b):
+    if a > b:
+        z = float(a - b)
+    else:
+        z = float(b) - 1 / a
+    return z - float(a)
+
+
+# Add rules for `float()`
+
+
+if __name__ == "__main__":
+    jt = compiler_pipeline(
+        example_2,
+        argtypes=(Int64, Int64),
+        ruleset=(
+            if_else_ruleset
+            | facts_function_types
+            | ruleset_type_infer_float  # < --- added for float()
+        ),
+        verbose=True,
+        converter_class=ConditionalExtendGraphtoRVSDG,
+        cost_model=MyCostModel(),
+        backend=Backend(),
+    )
+    args = (10, 33)
+    run_test(example_2, jt, args, verbose=True)
+    args = (7, 3)
+    run_test(example_2, jt, args, verbose=True)
+
+# ## Example 3: unify mismatching type
+#
+# What if type of `z` does not match across the branch?
+
+
+def example_3(a, b):
+    if a > b:
+        z = a - b  # this as int
+    else:
+        z = float(b) - 1 / a  # this is float
+    return z - float(a)
+
+
+# Add rules to signal error
+
+# ## Example 4: Improve error reporting
```

Comment:
Same as Example 3.  I don't think this changes at all relative to Ch 4, and we could skip it in this chapter.

Reply:
Fixed in latest commit